fission/src/systems/scene/GizmoSceneObject.ts

```diff
@@ -0,0 +1,194 @@
+            } else if (event.target.mode === "rotate") {
+                // scale on all axes
+                Array.from(World.SceneRenderer.sceneObjects.values())
+                    .filter(obj => obj instanceof GizmoSceneObject)
+                    .forEach(obj => {
+                        if (
+                            obj.gizmo.mode === "scale" &&
+                            event.target !== obj.gizmo &&
+                            obj.gizmo.object === event.target.object
+                        ) {
+                            obj.gizmo.dragging = false
+                            obj.gizmo.enabled = !event.value
+                            return
+                        }
+                    })
+            }
+        })
+    }
+
+    public Update(): void {
+        this._gizmo.updateMatrixWorld()
+
+        // updating the size of the gizmo based on the distance from the camera
+        const mainCameraFovRadians = (Math.PI * (this._mainCamera.fov * 0.5)) / 180
+        this._gizmo.setSize(
+            (this._size / this._mainCamera.position.distanceTo(this.gizmo.object!.position)) *
+                Math.tan(mainCameraFovRadians) *
+                1.9
+        )
+
+        /** Translating the obj changes to the mirabuf scene object */
+        if (this._parentObject) {
+            this._parentObject.DisablePhysics()
+            if (this.isDragging) {
+                this._parentObject.mirabufInstance.parser.rigidNodes.forEach(rn => {
+                    this.UpdateBodyPositionAndRotation(rn)
+                    this._parentObject?.UpdateNodeParts(rn, this.obj.matrix)
+                })
+            }
+        }
+
+        // creating enter key and escape key event listeners
+        if (InputSystem.isKeyPressed("Enter") && this._parentObject) {
+            // confirming placement of object
+            if (this._parentObject !== undefined) this._parentObject.EnablePhysics()
+            World.SceneRenderer.RemoveSceneObject(this.id)
+            return
+        } else if (InputSystem.isKeyPressed("Escape") && this._parentObject) {
+            // cancelling the creation of the mirabuf scene object
+            World.SceneRenderer.RemoveSceneObject(this._parentObject.id)
+            World.SceneRenderer.RemoveSceneObject(this.id)
+            return
+        }
+    }
+
+    public Dispose(): void {
+        this._gizmo.detach()
+        this._parentObject?.EnablePhysics()
+        World.SceneRenderer.RemoveObject(this._obj)
+        World.SceneRenderer.RemoveObject(this._gizmo)
+    }
+
+    /** changes the mode of the gizmo */
+    public SetMode(mode: GizmoMode) {
+        this._gizmo.setMode(mode)
+    }
+
+    /** updates body position and rotation for each body from the parent mirabuf */
+    public UpdateBodyPositionAndRotation(rn: RigidNodeReadOnly) {
+        if (!this._parentObject) return
+        World.PhysicsSystem.SetBodyPosition(
+            this._parentObject.mechanism.GetBodyByNodeId(rn.id)!,
```

Comment:
Are we certain that the result of `GetBodyByNodeId` will be defined? I'd definitely hope so, but would it be worth instead checking that value and then calling `World.PhysicsSystem.SetBodyPosition` only if it's defined?